### PR TITLE
Jvenstad/differentiate between deploy and developer keys

### PR DESCRIPTION
**Pull request by jonmv**
@bjorncs please review.

### CODE REVIEW

**Comment by jonmv** [comment]
@oyving @freva @ldalves FYI.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/filter/SignatureFilter.java [hunk 1]
```diff
@@ -47,23 +55,40 @@ public SignatureFilter(Controller controller) {
...
+                        if (tenant.isPresent() && tenant.get().developerKeys().containsKey(key)) {
+                            principal = tenant.get().developerKeys().get(key); // Precedence over headless user.
+                            roles.add(Role.reader(id.tenant()));
+                            roles.add(Role.developer(id.tenant()));
+                        }
+
+                        if (principal != null)
+                            securityContext = new SecurityContext(principal, roles);
+                    }
+                }
```
Comment (bjorncs) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/filter/SignatureFilter.java]: Consider moving this branch to a separate method to make the separation between developer and deploy keys more clear.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java [hunk 2]
```diff
@@ -81,7 +81,9 @@
               Policy.applicationUpdate,
               Policy.applicationDelete,
               Policy.applicationOperations,
-              Policy.developmentDeployment),
+              Policy.developmentDeployment,
+              Policy.keyManagement,
```
Comment (bjorncs) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java]: Deploy keys will now be granted all these permissions?

**Review by bjorncs — APPROVED** [review]
See comments

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 3]
```diff
@@ -377,8 +379,9 @@ private HttpResponse addDeveloperKey(String tenantName, HttpRequest request) {
 
         Principal user = request.getJDiscRequest().getUserPrincipal();
         String pemDeveloperKey = toSlime(request.getData()).get().field("key").asString();
+        PublicKey developerKey = KeyUtils.fromPemEncodedPublicKey(pemDeveloperKey);
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: The `Throwable.getMessage()` will be shown along with a 400 status code. Hopefully the message is good enough?

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/persistence/ApplicationSerializerTest.java [hunk 4]
```diff
@@ -64,6 +63,15 @@
...
     private static final ZoneId zone2 = ZoneId.from("prod", "us-east-3");
+    private static final PublicKey publicKey = KeyUtils.fromPemEncodedPublicKey("-----BEGIN PUBLIC KEY-----\n" +
+                                                                                "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEuKVFA8dXk43kVfYKzkUqhEY2rDT9\n" +
+                                                                                "z/4jKSTHwbYR8wdsOSrJGVEUPbS2nguIJ64OJH7gFnxM6sxUVj+Nm2HlXw==\n" +
+                                                                                "-----END PUBLIC KEY-----\n");
+    private static final PublicKey otherPublicKey = KeyUtils.fromPemEncodedPublicKey("-----BEGIN PUBLIC KEY-----\n" +
+                                                                                     "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEFELzPyinTfQ/sZnTmRp5E4Ve/sbE\n" +
+                                                                                     "pDhJeqczkyFcT2PysJ5sZwm7rKPEeXDOhzTPCyRvbUqc2SGdWbKUGGa/Yw==\n" +
+                                                                                     "-----END PUBLIC KEY-----\n");
+
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-server/src/test/java/com/yahoo/vespa/hosted/controller/persistence/ApplicationSerializerTest.java]: Ho ho, multi-line comments don't work.

[the remark this answers is not in the document]
Comment (jonmv) [reply controller-server/src/test/java/com/yahoo/vespa/hosted/controller/persistence/ApplicationSerializerTest.java]: Will keep that in mind!

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/user/UserApiTest.java [hunk 5]
```diff
@@ -23,6 +22,17 @@
...
 
     private static final String responseFiles = "src/test/java/com/yahoo/vespa/hosted/controller/restapi/user/responses/";
+    private static final String pemPublicKey = "-----BEGIN PUBLIC KEY-----\n" +
+                                               "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEuKVFA8dXk43kVfYKzkUqhEY2rDT9\n" +
+                                               "z/4jKSTHwbYR8wdsOSrJGVEUPbS2nguIJ64OJH7gFnxM6sxUVj+Nm2HlXw==\n" +
+                                               "-----END PUBLIC KEY-----\n";
+    private static final String otherPemPublicKey = "-----BEGIN PUBLIC KEY-----\n" +
+                                                    "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEFELzPyinTfQ/sZnTmRp5E4Ve/sbE\n" +
+                                                    "pDhJeqczkyFcT2PysJ5sZwm7rKPEeXDOhzTPCyRvbUqc2SGdWbKUGGa/Yw==\n" +
+                                                    "-----END PUBLIC KEY-----\n";
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/user/UserApiTest.java]: Some of the keys (like here) are also part of response files. But keeping hard-coded values to a minimum sounds like a good idae.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java (hunk 2 above)
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java]: Yes.

Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java]: Well, current deploy keys will be, for a while. Only (new) developer keys will have these privileges a few weeks from now.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/filter/SignatureFilter.java (hunk 1 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/filter/SignatureFilter.java]: Current code is better, but will consider this when I remove outdated code paths again in a couple of weeks.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bjorncs) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: You may want to prefix the message with "invalid public key" or similar. Error messages from the Java are often cryptic...

**Pull request merged**